libsolidity/analysis/ControlFlowAnalyzer.cpp

```diff
@@ -0,0 +1,129 @@
+				unassignedAtFunctionEntry.insert(returnParameter.get());
+	}
+
+	stack<CFGNode const*> nodesToTraverse;
+	nodesToTraverse.push(_functionEntry);
+
+	// walk all paths from entry with maximal set of unassigned return values
+	while (!nodesToTraverse.empty())
+	{
+		auto node = nodesToTraverse.top();
+		nodesToTraverse.pop();
+
+		auto& unassignedAtNode = unassigned[node];
+
+		if (node->block.returnStatement != nullptr)
+			if (node->block.returnStatement->expression())
+				unassignedAtNode.clear();
+		if (!unassignedAtNode.empty())
+		{
+			// kill all return values to which a value is assigned
+			for (auto expression: node->block.expressions)
+			{
+				if (auto const* assignment = dynamic_cast<Assignment const*>(expression))
+				{
+					stack<Expression const*> expressions;
+					expressions.push(&assignment->leftHandSide());
+					while (!expressions.empty())
+					{
+						Expression const* expression = expressions.top();
+						expressions.pop();
+
+						if (auto const *tuple = dynamic_cast<TupleExpression const*>(expression))
+							for (auto const& component: tuple->components())
+								expressions.push(component.get());
+						else if (auto const* identifier = dynamic_cast<Identifier const*>(expression))
+							if (auto const* variableDeclaration = dynamic_cast<VariableDeclaration const*>(
+								identifier->annotation().referencedDeclaration
+							))
+								unassignedAtNode.erase(variableDeclaration);
+					}
+				}
+			}
+			// kill all return values referenced in inline assembly
+			// a reference is enough, checking whether there actually was an assignment might be overkill
+			for (auto assembly: node->block.inlineAssemblyStatements)
+				for (auto const& ref: assembly->annotation().externalReferences)
+					if (auto variableDeclaration = dynamic_cast<VariableDeclaration const*>(ref.second.declaration))
+						unassignedAtNode.erase(variableDeclaration);
+		}
+
+		for (auto const& exit: node->exits)
+		{
+			auto& unassignedAtExit = unassigned[exit];
+			auto oldSize = unassignedAtExit.size();
+			unassignedAtExit.insert(unassignedAtNode.begin(), unassignedAtNode.end());
+			// traverse an exit, if we are on a path with new unassigned return values to consider
+			// this will terminate, since there is only a finite number of unassigned return values
+			if (unassignedAtExit.size() > oldSize)
+				nodesToTraverse.push(exit);
+		}
+	}
+
+	if (unassigned[_functionExit].size() > 0)
+	{
+		vector<VariableDeclaration const*> unassignedOrdered(
+			unassigned[_functionExit].begin(),
+			unassigned[_functionExit].end()
+		);
+		sort(
+			unassignedOrdered.begin(),
+			unassignedOrdered.end(),
+			[](VariableDeclaration const* lhs, VariableDeclaration const* rhs) -> bool {
+				return lhs->id() < rhs->id();
+			}
+		);
+		for (auto returnVal: unassignedOrdered)
+			m_errorReporter.warning(returnVal->location(), "uninitialized storage pointer may be returned");
```

Comment:
I don't think we can make this an error for 0.5.0 yet. There are still corner cases where the current analysis will result in false positives, e.g.:

```
contract C {
  function alwaysRevertWithSomeSpecificReason() {
    revert(...);
  }
  struct S { uint a; };
  S s;
  function f(bool flag) returns (S storage r) {
    if (flag) r = s;
    else alwaysRevertWithSomeSpecificReason();
  }
}
```
This example would require cross-function control flow, but even if we had that, I don't think it should be an error - consider:
```
contract C {
  struct S { uint a; };
  S s1;
  S s2;
  function f(bool a, bool b) returns (S storage r) {
    require(a || b);
    if (a) r = s1;
    if (b) r = s2;
  }
}
```
Using the SMT solver may help for such cases in the future (the CFG is rather dumb without it), but I'm not sure whether we can ever catch all false positives and at least for now I don't think it can be an error.

Reply:
We have the policy that every warning can be silenced by a simple code change. The obvious change here would be to use `r = r` in problematic code paths. This is even a stack assignment, so it has no practical influence on gas costs. So I think the false positives should not be an issue here.

What could be an argument is that everything that is an error has to be specified somehow, and it is probably very hard to formalize the control flow graph analysis.